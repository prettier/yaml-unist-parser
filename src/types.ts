import type * as YAML from "yaml";

export type ParsedCST = ReturnType<typeof YAML.parseCST>;

export interface Node {
  type: string;
  position: Position;
  /** @internal non-enumerable */
  _parent?: YamlUnistNode | null;
}

export interface Position {
  start: Point;
  end: Point;
}

export interface Range {
  origStart: number;
  origEnd: number;
}

export interface Point {
  /** 1-based */
  line: number;
  /** 1-based */
  column: number;
  /** 0-based */
  offset: number;
}

export interface Parent extends Node {
  children: Node[];
}

export interface Literal extends Node {
  value: string;
}

// -----------------------------------------------------------------------------

export interface Content {
  anchor: null | Anchor;
  tag: null | Tag;
  /** comments between the node and its tag/anchor */
  middleComments: Comment[];
}

export interface CommentAttachable
  extends LeadingCommentAttachable,
    TrailingCommentAttachable {}

export interface LeadingCommentAttachable {
  /** comments in front of the node */
  leadingComments: Comment[];
}

export interface TrailingCommentAttachable {
  /** comment on the same line of the node */
  trailingComment: null | Comment;
}

export interface EndCommentAttachable {
  /** comments after the node with greater column */
  endComments: Comment[];
}

export interface YAMLSyntaxError extends SyntaxError {
  source: string;
  position: Position;
}

export type YamlUnistNode =
  | Comment
  | Tag
  | Anchor
  | Root
  | Document
  | DocumentHead
  | DocumentBody
  | Directive
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | MappingItem
  | MappingKey
  | MappingValue
  | Sequence
  | SequenceItem
  | FlowMapping
  | FlowMappingItem
  | FlowSequence
  | FlowSequenceItem;

export type ContentNode = Extract<YamlUnistNode, Content>;

// -----------------------------------------------------------------------------

export interface Comment extends Literal {
  type: "comment";
}

export interface Anchor extends Literal {
  type: "anchor";
}

export interface Tag extends Literal {
  type: "tag";
}

export interface Root extends Parent {
  type: "root";
  children: Document[];
  comments: Comment[];
}

export interface Document extends Parent, TrailingCommentAttachable {
  type: "document";
  children: [DocumentHead, DocumentBody];
}

export interface DocumentHead
  extends Parent,
    EndCommentAttachable,
    TrailingCommentAttachable {
  type: "documentHead";
  children: Directive[];
}

export interface DocumentBody extends Parent, EndCommentAttachable {
  type: "documentBody";
  children: [] | [ContentNode];
}

export interface Directive extends Node, CommentAttachable {
  type: "directive";
  name: string;
  parameters: string[];
}

export interface Alias extends Literal, Content, CommentAttachable {
  type: "alias";
}

export interface BlockValue extends Literal, Content, LeadingCommentAttachable {
  chomping: "clip" | "keep" | "strip";
  indent: null | number;
  /** comment between indicator and the value */
  indicatorComment: null | Comment;
}

export interface BlockLiteral extends BlockValue {
  type: "blockLiteral";
}

export interface BlockFolded extends BlockValue {
  type: "blockFolded";
}

export interface Plain extends Literal, Content, CommentAttachable {
  type: "plain";
}

export interface QuoteValue extends Literal, Content, CommentAttachable {}

export interface QuoteSingle extends QuoteValue {
  type: "quoteSingle";
}

export interface QuoteDouble extends QuoteValue {
  type: "quoteDouble";
}

export interface Mapping extends Parent, Content, LeadingCommentAttachable {
  type: "mapping";
  children: MappingItem[];
}

export interface MappingItemBase extends Parent, LeadingCommentAttachable {
  /** key-value pair */
  children: [MappingKey, MappingValue];
}

export interface MappingItem extends MappingItemBase {
  type: "mappingItem";
}

export interface MappingKey
  extends Parent,
    TrailingCommentAttachable,
    EndCommentAttachable {
  type: "mappingKey";
  children: [] | [ContentNode];
}

export interface MappingValue
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingValue";
  children: [] | [ContentNode];
}

export interface Sequence
  extends Parent,
    Content,
    LeadingCommentAttachable,
    EndCommentAttachable {
  type: "sequence";
  children: SequenceItem[];
}

export interface SequenceItemBase extends Parent {
  children: [] | [ContentNode];
}

export interface SequenceItem
  extends SequenceItemBase,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequenceItem";
}

export interface FlowCollection
  extends Parent,
    Content,
    CommentAttachable,
    EndCommentAttachable {
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowMapping extends FlowCollection {
  type: "flowMapping";
  children: FlowMappingItem[];
}

export interface FlowMappingItem extends MappingItemBase {
  type: "flowMappingItem";
}

export interface FlowSequence extends FlowCollection {
  type: "flowSequence";
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowSequenceItem extends SequenceItemBase {
  type: "flowSequenceItem";
}
