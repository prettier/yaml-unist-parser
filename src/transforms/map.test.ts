import {
  getFirstContent,
  TestCase,
  testCases,
  testSyntaxError,
} from "../helpers.js";
import { Mapping, Root, Sequence } from "../types.js";

testSyntaxError("asd: 123\nqwe\n");

testCases([
  ...createTestCases(":   "),
  ...createTestCases(": 123  "),
  ...createTestCases("a:   "),
  ...createTestCases("a: 123  "),
  ...createTestCases("a: 123  \nb: 456  ", true),
  ...createTestCases("? a  \n: 123  \n? b  \n: 456  ", true),
  ...createTestCases("? abc\n? def", true),
  ...createTestCases("? 123"),
  ...createTestCases("def:     456 # hello"),
  [
    "!!map\n #123 \n&anchor # 456\na: 123",
    [
      getMapping(),
      root => getFirstContent<Mapping>(root).middleComments[0],
      root => getFirstContent<Mapping>(root).middleComments[1],
    ],
  ],
  [
    "x:\n  - &a\n    key1: value1\n  - &b\n    key2: value2\nfoo:\n  bar: baz\n  <<: *a\n  <<: *b",
    getMappingItem(1),
  ],
  [
    "merge:\n- &A { a: 1 }\n- &B { b: 2 }\n- <<: [ *A, *B ]",
    root => (getMappingValue(0)(root).children[0] as Sequence).children[2],
  ],
  ["a:\n  b:\n   #b\n #a\n", getFirstContent()],
  ["a: !!str", getFirstContent()],
  ["<<:\n  a: b", getFirstContent()],
]);

function createTestCases(text: string, hasSecondItem = false): TestCase[] {
  return !hasSecondItem
    ? [
        [text, getMapping()],
        [text, getMappingItem(0)],
        [text, getMappingKey(0)],
        [text, getMappingValue(0)],
      ]
    : [
        [text, getMapping()],
        [text, getMappingItem(0)],
        [text, getMappingItem(1)],
        [text, getMappingKey(0)],
        [text, getMappingKey(1)],
        [text, getMappingValue(0)],
        [text, getMappingValue(1)],
      ];
}

function getMapping() {
  return getFirstContent<Mapping>();
}

function getMappingItem(itemIndex: number) {
  return (root: Root) => getMapping()(root).children[itemIndex];
}

function getMappingKey(itemIndex: number) {
  return (root: Root) => getMappingItem(itemIndex)(root).children[0];
}

function getMappingValue(itemIndex: number) {
  return (root: Root) => getMappingItem(itemIndex)(root).children[1];
}
