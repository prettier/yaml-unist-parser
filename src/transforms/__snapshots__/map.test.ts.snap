// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`"!!map\\n #123 \\n&anchor # 456\\na: 123" 1`] = `
mapping (4:1 ~ 4:7)
1 | !!map¶
2 | ·#123·¶
3 | &anchor·#·456¶
4 | a:·123¶
  | ^^^^^^
<mapping>
  <tag value="tag:yaml.org,2002:map" />
  <anchor value="anchor" />
  <middleComment value="123 ">
  <middleComment value=" 456">
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"!!map\\n #123 \\n&anchor # 456\\na: 123" 2`] = `
comment (2:2 ~ 2:7)
1 | !!map¶
2 | ·#123·¶
  |  ^^^^^
3 | &anchor·#·456¶
4 | a:·123¶
<comment value="123 " />
`;

exports[`"!!map\\n #123 \\n&anchor # 456\\na: 123" 3`] = `
comment (3:9 ~ 3:14)
1 | !!map¶
2 | ·#123·¶
3 | &anchor·#·456¶
  |         ^^^^^
4 | a:·123¶
<comment value=" 456" />
`;

exports[`":   " 1`] = `
mapping (1:1 ~ 1:2)
1 | :···¶
  | ^
<mapping>
  <mappingItem>
    <mappingKey>

    </mappingKey>
    <mappingValue>

    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`":   " 2`] = `
mappingItem (1:1 ~ 1:2)
1 | :···¶
  | ^
<mappingItem>
  <mappingKey>

  </mappingKey>
  <mappingValue>

  </mappingValue>
</mappingItem>
`;

exports[`":   " 3`] = `
mappingKey (1:1 ~ 1:1)
1 | :···¶
  | ~
<mappingKey>

</mappingKey>
`;

exports[`":   " 4`] = `
mappingValue (1:1 ~ 1:2)
1 | :···¶
  | ^
<mappingValue>

</mappingValue>
`;

exports[`": 123  " 1`] = `
mapping (1:1 ~ 1:6)
1 | :·123··¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>

    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`": 123  " 2`] = `
mappingItem (1:1 ~ 1:6)
1 | :·123··¶
  | ^^^^^
<mappingItem>
  <mappingKey>

  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`": 123  " 3`] = `
mappingKey (1:1 ~ 1:1)
1 | :·123··¶
  | ~
<mappingKey>

</mappingKey>
`;

exports[`": 123  " 4`] = `
mappingValue (1:1 ~ 1:6)
1 | :·123··¶
  | ^^^^^
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"<<:\\n  a: b" 1`] = `
mapping (1:1 ~ 2:7)
1 | <<:¶
  | ^^^^
2 | ··a:·b¶
  | ^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="<<" />
    </mappingKey>
    <mappingValue>
      <mapping>
        <mappingItem>
          <mappingKey>
            <plain value="a" />
          </mappingKey>
          <mappingValue>
            <plain value="b" />
          </mappingValue>
        </mappingItem>
      </mapping>
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"? 123" 1`] = `
mapping (1:1 ~ 1:6)
1 | ?·123¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="123" />
    </mappingKey>
    <mappingValue>

    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"? 123" 2`] = `
mappingItem (1:1 ~ 1:6)
1 | ?·123¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <plain value="123" />
  </mappingKey>
  <mappingValue>

  </mappingValue>
</mappingItem>
`;

exports[`"? 123" 3`] = `
mappingKey (1:1 ~ 1:6)
1 | ?·123¶
  | ^^^^^
<mappingKey>
  <plain value="123" />
</mappingKey>
`;

exports[`"? 123" 4`] = `
mappingValue (1:6 ~ 1:6)
1 | ?·123¶
  |      ~
<mappingValue>

</mappingValue>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 1`] = `
mapping (1:1 ~ 4:6)
1 | ?·a··¶
  | ^^^^^^
2 | :·123··¶
  | ^^^^^^^^
3 | ?·b··¶
  | ^^^^^^
4 | :·456··¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
  <mappingItem>
    <mappingKey>
      <plain value="b" />
    </mappingKey>
    <mappingValue>
      <plain value="456" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 2`] = `
mappingItem (1:1 ~ 2:6)
1 | ?·a··¶
  | ^^^^^^
2 | :·123··¶
  | ^^^^^
3 | ?·b··¶
4 | :·456··¶
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 3`] = `
mappingItem (3:1 ~ 4:6)
1 | ?·a··¶
2 | :·123··¶
3 | ?·b··¶
  | ^^^^^^
4 | :·456··¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <plain value="b" />
  </mappingKey>
  <mappingValue>
    <plain value="456" />
  </mappingValue>
</mappingItem>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 4`] = `
mappingKey (1:1 ~ 1:4)
1 | ?·a··¶
  | ^^^
2 | :·123··¶
3 | ?·b··¶
4 | :·456··¶
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 5`] = `
mappingKey (3:1 ~ 3:4)
1 | ?·a··¶
2 | :·123··¶
3 | ?·b··¶
  | ^^^
4 | :·456··¶
<mappingKey>
  <plain value="b" />
</mappingKey>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 6`] = `
mappingValue (2:1 ~ 2:6)
1 | ?·a··¶
2 | :·123··¶
  | ^^^^^
3 | ?·b··¶
4 | :·456··¶
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"? a  \\n: 123  \\n? b  \\n: 456  " 7`] = `
mappingValue (4:1 ~ 4:6)
1 | ?·a··¶
2 | :·123··¶
3 | ?·b··¶
4 | :·456··¶
  | ^^^^^
<mappingValue>
  <plain value="456" />
</mappingValue>
`;

exports[`"? abc\\n? def" 1`] = `
mapping (1:1 ~ 2:6)
1 | ?·abc¶
  | ^^^^^^
2 | ?·def¶
  | ^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="abc" />
    </mappingKey>
    <mappingValue>

    </mappingValue>
  </mappingItem>
  <mappingItem>
    <mappingKey>
      <plain value="def" />
    </mappingKey>
    <mappingValue>

    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"? abc\\n? def" 2`] = `
mappingItem (1:1 ~ 1:6)
1 | ?·abc¶
  | ^^^^^
2 | ?·def¶
<mappingItem>
  <mappingKey>
    <plain value="abc" />
  </mappingKey>
  <mappingValue>

  </mappingValue>
</mappingItem>
`;

exports[`"? abc\\n? def" 3`] = `
mappingItem (2:1 ~ 2:6)
1 | ?·abc¶
2 | ?·def¶
  | ^^^^^
<mappingItem>
  <mappingKey>
    <plain value="def" />
  </mappingKey>
  <mappingValue>

  </mappingValue>
</mappingItem>
`;

exports[`"? abc\\n? def" 4`] = `
mappingKey (1:1 ~ 1:6)
1 | ?·abc¶
  | ^^^^^
2 | ?·def¶
<mappingKey>
  <plain value="abc" />
</mappingKey>
`;

exports[`"? abc\\n? def" 5`] = `
mappingKey (2:1 ~ 2:6)
1 | ?·abc¶
2 | ?·def¶
  | ^^^^^
<mappingKey>
  <plain value="def" />
</mappingKey>
`;

exports[`"? abc\\n? def" 6`] = `
mappingValue (1:6 ~ 1:6)
1 | ?·abc¶
  |      ~
2 | ?·def¶
<mappingValue>

</mappingValue>
`;

exports[`"? abc\\n? def" 7`] = `
mappingValue (2:6 ~ 2:6)
1 | ?·abc¶
2 | ?·def¶
  |      ~
<mappingValue>

</mappingValue>
`;

exports[`"a:   " 1`] = `
mapping (1:1 ~ 1:3)
1 | a:···¶
  | ^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>

    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a:   " 2`] = `
mappingItem (1:1 ~ 1:3)
1 | a:···¶
  | ^^
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>

  </mappingValue>
</mappingItem>
`;

exports[`"a:   " 3`] = `
mappingKey (1:1 ~ 1:2)
1 | a:···¶
  | ^
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"a:   " 4`] = `
mappingValue (1:2 ~ 1:3)
1 | a:···¶
  |  ^
<mappingValue>

</mappingValue>
`;

exports[`"a: !!str" 1`] = `
mapping (1:1 ~ 1:9)
1 | a:·!!str¶
  | ^^^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="">
        <tag value="tag:yaml.org,2002:str" />
      </plain>
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a: 123  " 1`] = `
mapping (1:1 ~ 1:7)
1 | a:·123··¶
  | ^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a: 123  " 2`] = `
mappingItem (1:1 ~ 1:7)
1 | a:·123··¶
  | ^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`"a: 123  " 3`] = `
mappingKey (1:1 ~ 1:2)
1 | a:·123··¶
  | ^
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"a: 123  " 4`] = `
mappingValue (1:2 ~ 1:7)
1 | a:·123··¶
  |  ^^^^^
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"a: 123  \\nb: 456  " 1`] = `
mapping (1:1 ~ 2:7)
1 | a:·123··¶
  | ^^^^^^^^^
2 | b:·456··¶
  | ^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <plain value="123" />
    </mappingValue>
  </mappingItem>
  <mappingItem>
    <mappingKey>
      <plain value="b" />
    </mappingKey>
    <mappingValue>
      <plain value="456" />
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"a: 123  \\nb: 456  " 2`] = `
mappingItem (1:1 ~ 1:7)
1 | a:·123··¶
  | ^^^^^^
2 | b:·456··¶
<mappingItem>
  <mappingKey>
    <plain value="a" />
  </mappingKey>
  <mappingValue>
    <plain value="123" />
  </mappingValue>
</mappingItem>
`;

exports[`"a: 123  \\nb: 456  " 3`] = `
mappingItem (2:1 ~ 2:7)
1 | a:·123··¶
2 | b:·456··¶
  | ^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="b" />
  </mappingKey>
  <mappingValue>
    <plain value="456" />
  </mappingValue>
</mappingItem>
`;

exports[`"a: 123  \\nb: 456  " 4`] = `
mappingKey (1:1 ~ 1:2)
1 | a:·123··¶
  | ^
2 | b:·456··¶
<mappingKey>
  <plain value="a" />
</mappingKey>
`;

exports[`"a: 123  \\nb: 456  " 5`] = `
mappingKey (2:1 ~ 2:2)
1 | a:·123··¶
2 | b:·456··¶
  | ^
<mappingKey>
  <plain value="b" />
</mappingKey>
`;

exports[`"a: 123  \\nb: 456  " 6`] = `
mappingValue (1:2 ~ 1:7)
1 | a:·123··¶
  |  ^^^^^
2 | b:·456··¶
<mappingValue>
  <plain value="123" />
</mappingValue>
`;

exports[`"a: 123  \\nb: 456  " 7`] = `
mappingValue (2:2 ~ 2:7)
1 | a:·123··¶
2 | b:·456··¶
  |  ^^^^^
<mappingValue>
  <plain value="456" />
</mappingValue>
`;

exports[`"a:\\n  b:\\n   #b\\n #a\\n" 1`] = `
mapping (1:1 ~ 4:4)
1 | a:¶
  | ^^^
2 | ··b:¶
  | ^^^^^
3 | ···#b¶
  | ^^^^^^
4 | ·#a¶
  | ^^^
5 | ¶
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="a" />
    </mappingKey>
    <mappingValue>
      <endComment value="a">
      <mapping>
        <mappingItem>
          <mappingKey>
            <plain value="b" />
          </mappingKey>
          <mappingValue>
            <endComment value="b">
          </mappingValue>
        </mappingItem>
      </mapping>
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"def:     456 # hello" 1`] = `
mapping (1:1 ~ 1:21)
1 | def:·····456·#·hello¶
  | ^^^^^^^^^^^^^^^^^^^^
<mapping>
  <mappingItem>
    <mappingKey>
      <plain value="def" />
    </mappingKey>
    <mappingValue>
      <plain value="456">
        <trailingComment value=" hello">
      </plain>
    </mappingValue>
  </mappingItem>
</mapping>
`;

exports[`"def:     456 # hello" 2`] = `
mappingItem (1:1 ~ 1:21)
1 | def:·····456·#·hello¶
  | ^^^^^^^^^^^^^^^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="def" />
  </mappingKey>
  <mappingValue>
    <plain value="456">
      <trailingComment value=" hello">
    </plain>
  </mappingValue>
</mappingItem>
`;

exports[`"def:     456 # hello" 3`] = `
mappingKey (1:1 ~ 1:4)
1 | def:·····456·#·hello¶
  | ^^^
<mappingKey>
  <plain value="def" />
</mappingKey>
`;

exports[`"def:     456 # hello" 4`] = `
mappingValue (1:4 ~ 1:21)
1 | def:·····456·#·hello¶
  |    ^^^^^^^^^^^^^^^^^
<mappingValue>
  <plain value="456">
    <trailingComment value=" hello">
  </plain>
</mappingValue>
`;

exports[`"merge:\\n- &A { a: 1 }\\n- &B { b: 2 }\\n- <<: [ *A, *B ]" 1`] = `
sequenceItem (4:1 ~ 4:17)
1 | merge:¶
2 | -·&A·{·a:·1·}¶
3 | -·&B·{·b:·2·}¶
4 | -·<<:·[·*A,·*B·]¶
  | ^^^^^^^^^^^^^^^^
<sequenceItem>
  <mapping>
    <mappingItem>
      <mappingKey>
        <plain value="<<" />
      </mappingKey>
      <mappingValue>
        <flowSequence>
          <flowSequenceItem>
            <alias value="A" />
          </flowSequenceItem>
          <flowSequenceItem>
            <alias value="B" />
          </flowSequenceItem>
        </flowSequence>
      </mappingValue>
    </mappingItem>
  </mapping>
</sequenceItem>
`;

exports[`"x:\\n  - &a\\n    key1: value1\\n  - &b\\n    key2: value2\\nfoo 1`] = `
mappingItem (6:1 ~ 9:9)
1 | x:¶
2 | ··-·&a¶
3 | ····key1:·value1¶
4 | ··-·&b¶
5 | ····key2:·value2¶
6 | foo:¶
  | ^^^^^
7 | ··bar:·baz¶
  | ^^^^^^^^^^^
8 | ··<<:·*a¶
  | ^^^^^^^^^
9 | ··<<:·*b¶
  | ^^^^^^^^
<mappingItem>
  <mappingKey>
    <plain value="foo" />
  </mappingKey>
  <mappingValue>
    <mapping>
      <mappingItem>
        <mappingKey>
          <plain value="bar" />
        </mappingKey>
        <mappingValue>
          <plain value="baz" />
        </mappingValue>
      </mappingItem>
      <mappingItem>
        <mappingKey>
          <plain value="<<" />
        </mappingKey>
        <mappingValue>
          <alias value="a" />
        </mappingValue>
      </mappingItem>
      <mappingItem>
        <mappingKey>
          <plain value="<<" />
        </mappingKey>
        <mappingValue>
          <alias value="b" />
        </mappingValue>
      </mappingItem>
    </mapping>
  </mappingValue>
</mappingItem>
`;

exports[`Implicit map keys need to be followed by map values 1`] = `
"Implicit map keys need to be followed by map values
1 | asd:·123¶
2 | qwe¶
  | ^^^
3 | ¶"
`;
